### PR TITLE
Add cocoaAction property to Action

### PR DESCRIPTION
**Pull request by neilpa**
As discussed in #1999

### CODE REVIEW

**Comment by rpowelll** [comment]
:+1: This seems like a better approach than making users manage `CocoaAction` instances themselves

File: ReactiveCocoa/Swift/Action.swift [hunk 1]
```diff
@@ -44,6 +44,16 @@ public final class Action<Input, Output, Error: ErrorType> {
 	/// Whether the instantiator of this action wants it to be enabled.
 	private let userEnabled: PropertyOf<Bool>
 
+	/// Lazy creation and storage of a UI bindable CocoaAction.
+	///
+	/// The default version force casts the AnyObject? input from CocoaAction to
+	/// match `Input` but this property can be overriden for alternative behavior.
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/Action.swift]: This property can't be overridden, since the class is `final`.

**Comment by jspahrsummers** [comment]
:+1: in general, but I'm nervous about the implicit (hidden) use of `as!`.

Can we create the `CocoaAction` with an overloaded `private` function that automatically handles at least the specific case of `Input == ()`? That would avoid the most common potential crash.

File: ReactiveCocoa/Swift/Action.swift (hunk 1 above)
Comment (neilpa) [inline ReactiveCocoa/Swift/Action.swift]: Poor wording, I meant re-assigned. However, this warning is mostly moot if the `private` function works.

**Comment by neilpa** [comment]
I'll give that a try

**Comment by neilpa** [comment]
I'm not sure the overloaded function approach is going to work. I tried multiple variations of the following but the wrong wrapper is always chose for `()` input. Is there some way to better discriminate these cases to effectively get `Input = ()` and `Input != ()`?

```swift
lazy var cocoaAction: CocoaAction = { _ in Action.wrapAction(self) }()

static func wrapAction(action: Action<(), Output, Error>) -> CocoaAction {
	return CocoaAction(action, input: ())
}

// Also tried `Action<Any, Output, Error>` here since that should cover all cases with above
// but then the compiler fails to infer types.
static func wrapAction(action: Action) -> CocoaAction {
	return CocoaAction(action) { $0 as! Input }
}
```

**Comment by jspahrsummers** [comment]
Maybe it'd be easier to have an overloaded function for transforming the input, `-> ()` or `-> T`.

**Comment by neilpa** [comment]
No go :sob:. I even tried this and found a compiler error I hadn't seen before.

screen shot 2015-06-16 at 12 02 40 am

**Comment by ikesyo** [comment]
Yep, `Void` is a empty (0 element) tuple, and tuples are not extendable...

**Comment by jspahrsummers** [comment]
Sorry, I meant we could have some functions like so:

```swift
static func transformCocoaActionInput(input: AnyObject?) -> () {
    return ()
}

static func transformCocoaActionInput(input: AnyObject?) -> T? {
    return input as? T
}
```

etc., as needed.

**Comment by neilpa** [comment]
Yea, I tried that first before trying the protocol (that's what my "No go" comment was referring to). The problem is it still requires a catch-all `AnyObject? -> Input` with force casting and the compiler always picks that over the specialized overloads.

**Comment by jspahrsummers** [comment]
In lieu of a perfect solution, I guess, maybe we can call this `unsafeCocoaAction` or something like that?

**Comment by neilpa** [comment]
I'd be fine with that.

On a related note, I was experimenting last week with some [UIKit bindings](https://github.com/neilpa/Rex/pull/15) that would obviate the need for this property.

```swift
extension UIButton {
    public var rac_pressed: MutableProperty<CocoaAction> { ... }
}

class MyView: UIView {
    let model: MutableProperty<MyModel>
    let button: UIButton!

    public func viewDidLoad() {
        button.rac_pressed <~ model.producer
            |> map { CocoaAction($0.someAction, input: () }
    }
}

**Comment by neilpa** [comment]
Updated to `unsafeCocoaAction`

**Comment by jspahrsummers** [comment]
:shipit:

**Comment by almassapargali** [comment]
I'm rewriting some parts of my app because I found memory leaks lately. And now I'm thinking, won't current implementation of `unsafeCocoaAction` lead to memory leak, since `Action` stores `CocoaAction`, who retains `Action` in it's `_execute` closure?

**Comment by guidomb** [comment]
What is the "best" way to define an action in a view model that does not care about the input? Using `AnyObject?` or `Void` generates a runtime error when using the `unsafeCocoaAction` property

**Comment by almassapargali** [comment]
Yes. Because unsafeCocoaAction force casts to Input type (that's why it's unsafe). If your action takes Void as Input you may assign unsafeCocoaAction property with your own CocoaAction. Like:

myAction.unsafeCocoaAction = { [unowned myAction] in
    return CocoaAction(myAction, input: ())
}()

**Comment by guidomb** [comment]
@almassapargali Thanks! Looks like is more typing than just using `AnyObject!` as the Action's input type.

**Pull request merged**